src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -432,17 +440,24 @@ private void AddDirectoryWatchUnlocked(WatchedDirectory? parent, string director
                 // asked for subdirectories to be included.
                 if (isNewDirectory && _includeSubdirectories)
                 {
-                    // This method is recursive.  If we expect to see hierarchies
-                    // so deep that it would cause us to overflow the stack, we could
-                    // consider using an explicit stack object rather than recursion.
-                    // This is unlikely, however, given typical directory names
-                    // and max path limits.
-                    foreach (string subDir in Directory.EnumerateDirectories(fullPath))
+                    try
```

Comment:
In #57139 the failing method is called by `Directory.EnumerateDirectories`, but the new try-catch block is also covering `AddDirectoryWatchUnlocked` so in theory if there is a new bug in `AddDirectoryWatchUnlocked` and it starts throwing, we won't notice that. Perhaps we could do sth like this:

```cs
using IEnumerator<string> enumerator = Directory.EnumerateDirectories(fullPath);
bool canMove;
do
{
    canMove = false;
    try
    {
        canMove = enumerator.MoveNext();
    }
    catch (DirectoryNotFoundException) { }
    catch (IOException ex) when (ex.HResult == Interop.Error.ENOTDIR.Info().RawErrno) { }

    if (canMove)
    {
        AddDirectoryWatchUnlocked(directoryEntry, System.IO.Path.GetFileName(enumerator.Current));
    }

} while (canMove);
```